subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClassLoaderUtils.java

```diff
@@ -51,4 +91,16 @@ public static void disableUrlConnectionCaching() {
             throw UncheckedException.throwAsUncheckedException(e);
         }
     }
+
+    public static JavaMethod<ClassLoader, Package[]> getPackagesMethod() {
+        return GET_PACKAGES_METHOD;
+    }
+
+    public static JavaMethod<ClassLoader, Package> getPackageMethod() {
+        return GET_PACKAGE_METHOD;
+    }
+
+    public static <T> Class<T> define(ClassLoader targetClassLoader, String className, byte[] clazzBytes) {
+        return Cast.uncheckedCast(UNSAFE.defineClass(className, clazzBytes, 0, clazzBytes.length, targetClassLoader, null));
```

Comment:
Branch build is running on CI. I checked on IBM JDK 8 locally, and the build passed. Note that this implementation should be slightly faster, because using `Unsafe` will bypass the security checks. However, it's not a perfect solution. In particular, JDK 9 comes with an official alternative to calling this method, which is `MethodHandles.Lookup#defineClass`. The javadocs for this new method says:

> Defines a class to the same class loader and in the same runtime package and protection domain as this lookup's lookup class.

Which is nice, because it's actually everything we do. The problem, however, is that I fail to see how to get a `Lookup` object for the decorated class. It has to be in the same package, so unless the `Lookup` object is created from the class itself, or a class in the same package, I fail to see how we can use this API. I'm going to check with some friends if they have ideas around this.

In theory, it means that we're safe for JDK 9, but that it would break with JDK 10.